dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,219 @@
+            # Logging as error as this should not happen
+            logger.error("Got multiple %s integrations !", self.dns_record_requirer.relation_name)
+            raise
+
+    def dns_record_relations_data_to_entries(
+        self,
+        relation_data: list[
+            tuple[dns_record.DNSRecordRequirerData, dns_record.DNSRecordProviderData]
+        ],
+    ) -> list[models.DnsEntry]:
+        """Convert DNSRecordRequirerData to a list of DnsEntry.
+
+        Args:
+            relation_data: input relation data
+        Returns:
+            A list of DnsEntry
+        """
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(models.create_dns_entry_from_requirer_entry(entry))
+        return entries
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle changed configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring workload")
+        self.dns_policy.configure(
+            dns_policy.DnsPolicyConfig.model_validate(
+                {
+                    "debug": "true" if self.config["debug"] else "false",
+                    "allowed-hosts": json.dumps(
+                        [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                    ),
+                }
+            )
+        )
+
+    def _on_start(self, _: ops.StartEvent) -> None:
+        """Handle start event."""
+
+    def _on_install(self, _: ops.InstallEvent) -> None:
+        """Handle install event."""
+        self.unit.status = ops.MaintenanceStatus("Preparing dns-policy-app")
+        self.dns_policy.setup(self.unit.name)
+        self._timer.start(
+            self.unit.name,
+            "reconcile",
+            constants.RECONCILE_TIMER_TIMEOUT,
+            constants.RECONCILE_TIMER_INTERVAL,
+        )
+        logger.info(
+            "Started reconcile timer at %s, interval of %s",
+            datetime.datetime.now(),
+            constants.RECONCILE_TIMER_INTERVAL,
+        )
+
+    def _on_database_created(self, _: DatabaseCreatedEvent) -> None:
+        """Handle database created.
+
+        Args:
+            event: Event triggering the database created handler.
+        """
+        self._handle_database_endpoint_changes()
+
+    def _on_database_endpoints_changed(self, _: DatabaseEndpointsChangedEvent) -> None:
+        """Handle endpoints change.
+
+        Args:
+            event: Event triggering the endpoints changed handler.
+        """
+        self._handle_database_endpoint_changes()
+
+    def _handle_database_endpoint_changes(self) -> None:
+        database_relation_data = self._database.get_relation_data()
+        self.unit.status = ops.MaintenanceStatus("Preparing database")
+        self.dns_policy.configure(
+            dns_policy.DnsPolicyConfig.model_validate(
```

Comment:
Greg: Let's create a function that creates the DnsPolicyConfig based on the db relation data and self.config.
Take example on this: https://github.com/canonical/digest-squid-auth-helper/blob/main/src/charm_state.py#L61-L114

Reply:
`from_charm()` function implemented